src/System.Linq/src/System/Linq/Enumerable.cs

```diff
@@ -907,14 +923,122 @@ public static IEnumerable<TSource> Skip<TSource>(this IEnumerable<TSource> sourc
+                    ? (IPartition<TSource>)new EmptyPartition<TSource>()
+                    : new SkipListIterator<TSource>(_source, minIndex, _maxIndex);
+            }
+
+            public IPartition<TSource> Take(int count)
+            {
+                int maxIndex = _minIndex + count - 1;
+                return new SkipListIterator<TSource>(_source, _minIndex, maxIndex >= _maxIndex ? _maxIndex : maxIndex);
+            }
+
+            public TSource ElementAt(int index)
+            {
+                if ((uint)index > (uint)_maxIndex - _minIndex || index >= _source.Count - _minIndex) throw Error.ArgumentOutOfRange("index");
+                return _source[_minIndex + index];
+            }
+
+            public TSource ElementAtOrDefault(int index)
+            {
+                return (uint)index > (uint)_maxIndex - _minIndex || index >= _source.Count - _minIndex ? default(TSource) : _source[_minIndex + index];
+            }
+
+            public TSource First()
+            {
+                if (_source.Count <= _minIndex) throw Error.NoElements();
+                return _source[_minIndex];
+            }
+
+            public TSource FirstOrDefault()
+            {
+                return _source.Count <= _minIndex ? default(TSource) : _source[_minIndex];
+            }
+
+            public TSource Last()
+            {
+                int lastIndex = _source.Count - 1;
+                if (lastIndex < _minIndex) throw Error.NoElements();
+                return _source[lastIndex > _maxIndex ? _maxIndex : lastIndex];
+            }
+
+            public TSource LastOrDefault()
+            {
+                int lastIndex = _source.Count - 1;
+                return lastIndex < _minIndex ? default(TSource) : _source[lastIndex > _maxIndex ? _maxIndex : lastIndex];
+            }
+
+            public TSource[] ToArray()
+            {
+                int count = Count;
+                if (count == 0) return Array.Empty<TSource>();
+                TSource[] array = new TSource[count];
+                for (int i = 0, curIdx = _minIndex; i != array.Length; ++i, ++curIdx)
+                    array[i] = _source[curIdx];
+                return array;
+            }
+
+            public List<TSource> ToList()
+            {
+                int count = Count;
+                if (count == 0) return new List<TSource>();
+                List<TSource> list = new List<TSource>(count);
+                int end = _minIndex + count;
+                for (int i = _minIndex; i != end; ++i)
+                    list.Add(_source[i]);
+                return list;
+            }
+
+            public int Count
             {
-                yield return source[count++];
+                get
+                {
+                    int count = _source.Count;
+                    if (count <= _minIndex) return 0;
+                    return (count >= _maxIndex ? _maxIndex + 1 : count) - _minIndex;
```

Comment:
I don't think this is right.  Let's say someone did Take(6) on a list of 5 items.  From what I see in the code, that'll make _maxIndex == 5, count == 5, and _minIndex == 0.  Then count will be equal to _maxIndex, so we'll return _maxIndex + 1 - _minIndex, which is 6, but we should be returning 5.